File: src/examples/protect-endpoint/middleware.java
package com.example.kessel;

import org.springframework.web.bind.annotation.*;
import org.springframework.http.ResponseEntity;
import org.springframework.http.HttpStatus;
import jakarta.servlet.*;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import java.io.IOException;

import org.project_kessel.api.inventory.v1beta2.*;
import org.project_kessel.api.rbac.v2.Utils;
import static org.project_kessel.api.inventory.v1beta2.KesselInventoryServiceGrpc.KesselInventoryServiceBlockingStub;

//#region setup
public class ProtectEndpointExample {
    private final KesselInventoryServiceBlockingStub kesselClient;

    public ProtectEndpointExample() {
        this.kesselClient = new ClientBuilder("localhost:9000")
            .insecure()
            .build();
    }
//#endregion

//#region middleware
    // Servlet filter for permission checking
    public class PermissionFilter implements Filter {
        private final String relation;

        public PermissionFilter(String relation) {
            this.relation = relation;
        }

        @Override
        public void doFilter(ServletRequest request, ServletResponse response, FilterChain chain)
                throws IOException, ServletException {
            HttpServletRequest httpRequest = (HttpServletRequest) request;
            HttpServletResponse httpResponse = (HttpServletResponse) response;

            String userId = httpRequest.getHeader("X-User-ID");
            String workspaceId = httpRequest.getHeader("X-Workspace-ID");

            if (userId == null || workspaceId == null) {
                httpResponse.sendError(HttpStatus.BAD_REQUEST.value(), "Missing required headers");
                return;
            }

            CheckRequest checkRequest = CheckRequest.newBuilder()
                .setObject(Utils.workspaceResource(workspaceId))
                .setRelation(relation)
                .setSubject(Utils.principalSubject(userId, "redhat"))
                .build();

            try {
                CheckResponse checkResponse = kesselClient.check(checkRequest);

                if (checkResponse.getAllowed() != Allowed.ALLOWED_TRUE) {
                    httpResponse.sendError(HttpStatus.FORBIDDEN.value(), "Forbidden");
                    return;
                }

                chain.doFilter(request, response);
            } catch (Exception e) {
                httpResponse.sendError(HttpStatus.INTERNAL_SERVER_ERROR.value(),
                    "Permission check failed");
            }
        }
    }
//#endregion

//#region usage
    @RestController
    @RequestMapping("/integrations")
    public class IntegrationController {

        @GetMapping("/{id}")
        public ResponseEntity<?> getIntegration(@PathVariable String id) {
            return ResponseEntity.ok(Map.of("integration_id", id));
        }

        @PutMapping("/{id}")
        public ResponseEntity<?> updateIntegration(@PathVariable String id) {
            return ResponseEntity.ok(Map.of("status", "updated"));
        }
    }
//#endregion
}

File: src/examples/protect-endpoint/middleware.rb
require 'sinatra'
require 'kessel/inventory/v1beta2/client_builder'
require 'kessel/inventory/v1beta2/check_request_pb'
require 'kessel/inventory/v1beta2/allowed_pb'
require 'kessel/rbac/v2'

include Kessel::RBAC::V2

# region setup
# Initialize Kessel client
client = Kessel::Inventory::V1beta2::ClientBuilder
  .new("localhost:9000")
  .insecure
  .build
# endregion

# region middleware
# Helper to check permissions
def check_permission(client, relation)
  user_id = request.env['HTTP_X_USER_ID']
  workspace_id = request.env['HTTP_X_WORKSPACE_ID']

  halt 400, { error: 'Missing required headers' }.to_json if user_id.nil? || workspace_id.nil?

  check_request = Kessel::Inventory::V1beta2::CheckRequest.new(
    object: workspace_resource(workspace_id),
    relation: relation,
    subject: principal_subject(user_id, "redhat")
  )

  begin
    response = client.check(check_request)
    unless response.allowed == Kessel::Inventory::V1beta2::Allowed::ALLOWED_TRUE
      halt 403, { error: 'Forbidden' }.to_json
    end
  rescue StandardError => e
    logger.error "Check failed: #{e.message}"
    halt 500, { error: 'Permission check failed' }.to_json
  end
end
# endregion

# region usage
get '/integrations/:id' do
  check_permission(client, "myservice_integration_view")
  { integration_id: params[:id] }.to_json
end

put '/integrations/:id' do
  check_permission(client, "myservice_integration_edit")
  { status: 'updated' }.to_json
end
# endregion
